### ISSUE TITLE
[Bug]: Don't correlate accounts with RPC endpoint.

### ISSUE THREAD

**Issue by MicahZoltu** [MetaMask/metamask-extension#15169]
### Describe the bug

When first unlocking the wallet, MetaMask sends a batch balance request for every account to the currently selected JSON-RPC endpoint.  This means that the JSON-RPC endpoint can link all of the user's accounts.  This also means that if a user *accidentally* connects to the wrong JSON-RPC endpoint it also leaks all of their addresses (not just the selected one) to that endpoint.

This is true also for other chains, as a user connecting to a test network or L2 via MM will also send the RPC provider for that chain all of their accounts rather than just the selected account.

### Steps to reproduce

1. Create a wallet with multiple accounts.
2. Sign-in to MM.
3. Notice that MeteaMask sends an `eth_call` that includes every account, not just the selected one, to the selected JSON-RPC provider.

### Version

10.16.2

### Browser

Brave

### Operating system

Windows

**Comment by LefterisJP** [comment]
Let me add a proverbial +1 here

**Comment by rfikki** [comment]
I also want to add a +1 here

**Comment by banteg** [comment]
on unlock, metamask queries `balances(address[],address[])` method of `0xb1f8e55c7f64d203c1400b9d8555d050f94adf39` and passes all your addresses as the first argument. since infura has full visibility into rpc requests, they can easily establish a link between all your addresses, as well as link all your addresses to your ip address and browser fingerprint, before you even make any action with the wallet.

how to verify for yourself:
1. open the browser extension settings
2. enable developer mode
3. click "inspect views" for metamask
4. open "network" tab, type "infura" in the search bar
5. unlock metamask
6. find an `eth_call` request to the address mentioned above

**Comment by LefterisJP** [comment]
Spoke with some team members in Twitter about it. The reason for this is speed and caching. So cache all data/balances for all user addresses and show them fast instead of querying only when "connecting".

I understand that reason from a product and user PoV. But this should be configurable. Whether we want this cache to be done or not.

As an expert user, focused on privacy I would love for the default to be **not doing** that. But I can see why you would default to doing it when your users complain about speed. Still I would opt for user education on privacy rather than speed here.

**Comment by ruhaise** [comment]
I think the default option should be not doing the current behavior or it has to be adequately communicated to the user.
so a +1 from me here.

**Comment by banteg** [comment]
@LefterisJP the batch querying would've been fine with a local client, but with one entity controlling so much of the stack (wallet+provider), this begs more scrutiny.

**Comment by paulmillr** [comment]
Seems like MM sends stuff to infura even before terms acceptance. Which contradicts what's being said in privacy policy.

This requires a more thorough approach: reviewing *all* network and infura calls and adding a selector that allows to select a node *before* wallet is initialized.

P.s. Hello all the usual folks ITT.

**Comment by danfinlay** [comment]
Hey Micah, just curious what you would hope to see here.

We want to be able to show a user's whole portfolio responsively (like the balances when you open the accounts menu). This means we need to query the RPC node for balances.

Separating those requests into separate RPC requests does not prevent a provider from seeing the IP/address combination. A user could also choose to spam the balance-querying endpoint with fake accounts, there's no way the provider can prove that those accounts belong to a user even if it was analyzing that data (which Infura is not, btw). I guess it could wait for signed payloads and then have more confidence in those.

Anyways: We don't do that, we aren't starting to do that, and even if we were doing that, splitting the analysis into separate requests wouldn't add any privacy.

So, what would you hope to see here?

**Comment by stoooops** [comment]
I think some of the 30M users would prefer privacy over optimizing for balance refresh.

A setting to de-batch the requests and send them lazily (allowing me to switch VPNs if I choose) might be a path forward.

**Comment by paulmillr** [comment]
@danfinlay there should be a network kill switch with 3 options.

1. Disable all network requests
2. Only enable requests to particular domains and block reqs to everything else.
3. Enable all network requests. 

There should also be full transparency as to when requests are done and why. Ideally an option that notifies user about every request, with its url.

**Comment by mohsenghajar** [comment]
@danfinlay I am not sure most people treat that as "portfolio" as each account can contain any number of different assets.  Most of us have different accounts for a reason.  But I agree it can't prove ownership; e.g. if I'm maintaining an account for my mom 
 and dad.

**Comment by paulmillr** [comment]
@danfinlay Regarding your Twitter points that you've "created the current architecture, it was worse before": sure, and you have respect for that.

However, at some point, someone in some bad country will go to jail just because you're having bad UX here.

This is no joke. Dive into it further and you'll see why some community members are adamant with their position.

**Comment by danfinlay** [comment]
> A setting to de-batch the requests and send them lazily or via user click (allowing me to switch VPNs if I choose) might be a path forward.

I like this as an option also because it allows us to reduce server costs, which also makes self-hosting or shared hosting cheaper for users. I'll note it as something to add to our planned privacy onboarding. I don't think it solves the issue, but it does make the correlation somewhat weaker. Captured as  #16694.

<img width="401" alt="Screen Shot 2022-11-25 at 10 39 49 AM" src="https://user-images.githubusercontent.com/542863/204041621-c5e00fca-ea87-4e24-bbcf-38c5ebb80b90.png">

> there should be a network kill switch with 3 options.
>  Disable all network requests

I agree, capturing.

>   Only enable requests to particular domains and block reqs to everything else.

I like this, and it seems like it should be easy enough to add since our permissions system already restricts many things by default. Could break some dapps. Anyways, captured as #16695.

>  Enable all network requests.

Right, the current behavior will probably still be available for people who want that "very portfolio, web2"-like experience.

> I am not sure most people treat that as "portfolio" as each account can contain any number of different assets. Most of us have different accounts for a reason.

We do a lot of user research. I am like you on this point, and I think we should serve that use case well, but it does seem like a majority of people just want to see everything they have.

**Comment by joshfraser** [comment]
Are you able to share any information on your data retention policies? How many government requests for PII data have you received and how did you respond?

**Comment by MicahZoltu** [comment]
TL;DR: We should strive to both:
1. Provide privacy protecting products and configurations, even if that means making sacrifices to some features.
2. Make sure users are very aware anytime they do some new action that may harm their privacy.

> We want to be able to show a user's whole portfolio responsively (like the balances when you open the accounts menu). This means we need to query the RPC node for balances.
> 
> Separating those requests into separate RPC requests does not prevent a provider from seeing the IP/address combination. A user could also choose to spam the balance-querying endpoint with fake accounts, there's no way the provider can prove that those accounts belong to a user even if it was analyzing that data (which Infura is not, btw). I guess it could wait for signed payloads and then have more confidence in those.
> 
> Anyways: We don't do that, we aren't starting to do that, and even if we were doing that, splitting the analysis into separate requests wouldn't add any privacy.
> 
> So, what would you hope to see here?

I believe that MetaMask users currently operate under the assumption that if they have multiple accounts with one active, only the one active account will be sent anywhere and their accounts are otherwise *reasonably* separated from each other.  While we cannot make strong guarantees to users that there will be no way for a powerful actor to correlate their accounts, I think that MetaMask should at least not make that trivial, especially for users who follow generally good OpSec practices like using a VPN.

I believe there are two potential paths to addressing this issue:
1. Make it very clear to users that any account they add to MetaMask will be tightly connected to all of the other added accounts.  This will make it so users understand the privacy tradeoff they are making when using MetaMask with multiple accounts.
3. Make it so MetaMask is defaulting to a "privacy first" policy, where sacrifices are made on features in exchange for privacy, and allow users to enable certain features (like bulk balance updates) only after acknowledging that they understand the privacy risks.

In an ideal world, anytime a user's privacy is harmed in exchange for some feature it would be purely opt-in and never the default, and there will always be a way to use a slightly less feature rich version of the product with full privacy.  In a less ideal but still better-than-today world users are at least fully aware of the privacy sacrifices they are making when using any particular tool (this doesn't mean buried in a 100 page privacy policy but rather presented to the user in a human friendly way the first time they take a privacy harming action like adding a second account).

The problem with the current situation is that (as seen on Twitter) even the most advanced users of MetaMask didn't realize that MetaMask was strongly correlating all of their addresses together.  We can be pretty confident that less sophisticated users were even less aware.

**Comment by MicahZoltu** [comment]
> We do a lot of user research. I am like you on this point, and I think we should serve that use case well, but it does seem like a majority of people just want to see everything they have.

I'm curious, in your user research are users made aware of the privacy tradeoffs they are making when asked whether they like feature A or feature B?  Or are they only shown the benefits of B over A, while the downsides are hidden from them unless they think *really* hard about it?

**Comment by danfinlay** [comment]
> Are you able to share any information on your data retention policies? How many government requests for PII data have you received and how did you respond?

We're in the midst of a data retention audit right now, if it results in information like this we can/will share, we will. I don't know the answers to this offhand.

> TL;DR: We should strive to both

We actually have worked to embody this. You may have noticed we rolled out a "portfolio view", which relies on some downright centralized data sources, like indexers and historical price feeds. Rather than building this feature set into the wallet by default, the user has an opportunity to link out to that dapp, and opt into what accounts to connect to it.

I consider this a strong distinction because MetaMask core allows the user to swap out the provider for one they trust. I think this is a significant distinction, and it means statements like like

> any account they add to MetaMask will be tightly connected to all of the other added accounts

are not strictly true. It depends on many things. Like:
- Whether they trust their RPC provider
- Whether they believe an RPC provider will assume all accounts queried for a balance are theirs

I think we could:
- look at more manually-triggered balance reloading, which seems like a good "maximally private even when on a public RPC" (#16694)
- Better discuss during onboarding what the privacy model of the RPC is.

> in your user research are users made aware of the privacy tradeoffs they are making when asked whether they like feature A or feature B?

We definitely strive to make sure users are opting into security/privacy tradeoffs.

I'm not sure "fearing the RPC will correlate your accounts" has been part of those discussions, as we've treated the RPC as a trusted data source, which the user can opt into for their own. I'm sure we'll be discussing whether that risk has been adequately disclaimed to the user more after this week.

Again, we have isolated the most web2-dependent portfolio-view type features into a separate dapp to ensure users are making consensual privacy acts, so it should be obvious that we do in fact care about helping users make informed privacy choices.

We'll continue to evaluate new perceived threats (like this "batched account rpc correlation attack"), and prioritize the things we believe are the biggest threats to our users at any time. I think we've recently often felt that users have had bigger risks from actual thieves/phishers/scammers, and so we have had more focus on reducing that incident rate over some of these more theoretical privacy attacks, but I think this week is a nice opportunity to address our top outstanding privacy issues.

**Comment by MicahZoltu** [comment]
Currently, users cannot *functionally* "just use their own RPC" to avoid data leaking.  This is because MetaMask doesn't allow setting the RPC until *after* account setup is complete, and you cannot remove the default Infura endpoints which makes it very easy to accidentally leak to Infura, and (probably the worst of it), if a dapp prompts MetaMask to switch to chain 1 it will switch to Infura *instead of* the custom RPC added for chain 1.

----

Note: The below is intended as constructive but critical feedback, not an attack.  I *do* suspect you all are trying to do good, but it seems there is a disconnect here between what you think you are doing and what everyone else thinks you are doing and perhaps you are not aware of that disconnect.

From a public perception standpoint, I think a big part of the reason a lot of the people I speak with (not counting Twitter here) think MetaMask isn't really user privacy aligned/focused is because issues like this sit for many months untouched, without any comment.  I suspect this issue would have continued to sit for many more months, possibly years (based on personal experience filing/following issues on this repository) without any further comment.  The other open issue I have right now I argue is an actually much worse security hole because it is leaking secrets (not private keys, but user secrets none the less) and it has similarly sat untouched for months (and again, based on my past experience probably many more months/year if it weren't for the current Twitter outrage).

Similarly, while I *want* to believe you that you care a lot about user security and protecting their funds, issues like #9437 sat open for 6 months when it should have (IMO) been triaged with just a hard failure (no UI needed, crash the app if necessary) within days of being noticed given the sheer volume of funds lost.  The fix then somehow disappeared earlier this year and once again took a while (I forget how long) before it was fixed again (which presumably should have been easy given the designs and implementation were already previously done and just got lost).

This is a recurring pattern where security/privacy auditors will report things here or to Hackerone and they go unfixed for very long periods of time without any feedback or often even acknowledgement that it is a problem.

I don't have a good answer on how to repair this perception, but I think a start would be to be more proactive about publicly acknowledging privacy/security issues soon after they are reported and providing explanations on why they aren't being addressed quickly or at all.  If you wait for a Twitter flash mob before commenting, then it just looks like you are doing PR damage control (common for large companies), not that you actually care (despite that I suspect you do care).

**Comment by danfinlay** [comment]
> Currently, users cannot functionally "just use their own RPC" to avoid data leaking. This is because MetaMask doesn't allow setting the RPC until after account setup is complete

We're already addressing this, and I noted it in #16696, so this point is now off topic.

> it seems there is a disconnect here between what you think you are doing and what everyone else thinks you are doing and perhaps you are not aware of that disconnect.

I'd counter that people are doing an extremely good job of ensuring we know exactly what they think of us!

> I think a big part of the reason a lot of the people I speak with (not counting Twitter here) think MetaMask isn't really user privacy aligned/focused is because issues like this sit for many months untouched, without any comment.

Ok, but I assure you this has much more to do with the fact that we continue to balance an impossible number of priorities. Also, the RPC has historically been treated as a trusted data provider, so I think it was reasonable to triage this issue in particular as low impact. Just because it is fashionable to distrust Infura this week doesn't mean it's going to be effort efficient to rework the entire application around this fringe desire.

> The other open issue I have right now I argue is an actually much worse security hole

Link?

> Similarly, while I want to believe you that you care a lot about user security and protecting their funds, issues like https://github.com/MetaMask/metamask-extension/issues/9437 sat open for 6 months when it should have (IMO) been triaged with just a hard failure

There had been a regression in that validation, and that issue did sit open far too long. That's a failure. To believe this proves some "not caring about protecting user funds" requires assuming total omniscience and some kind of cryptic motivation to help users make mistakes. I don't even know what you're accusing us of here. Mistakes happen. We do retros when they do and try to ensure that type of mistake can't happen again. It's bizarre to suggest an occasional mis-prioritization proves anything malicious.

> be more proactive about publicly acknowledging privacy/security issues soon after they are reported and providing explanations on why they aren't being addressed quickly or at all

The problem is awareness. It's not like we're ignoring known issues, it's that we're always managing a queue of high priority items, and so what you see as high priority does not necessarily match what we're even aware of as important. We've made continuous improvements to our division of responsibilities and triaging, but there's no way it's as good as it deserves to be given the type of product this is.

Btw, I would love for us to have a more liquid prioritization system. [I wrote some ideas for one here](https://roamresearch.com/#/app/capabul/page/5oxgola9b), and some of those ideas were built into [TokenLog](https://github.com/wslyvh/tokenlog). I think prioritization is important, maybe the most important thing, and I very much wish it was easier for more members of the community to "thumb the scale" when something was getting neglected for whatever reason. Of course, to build that ourselves we'd have to neglect other obviously important things, and so we go without many nice things. Open invitation!
